// @backstage/plugin-catalog-react/dist/alpha.esm.js
import { parseEntityRef, getCompoundEntityRef, stringifyEntityRef, RELATION_MEMBER_OF, RELATION_OWNED_BY } from '@backstage/catalog-model';
import { usePermission } from '@backstage/plugin-permission-react';
import '@backstage/core-plugin-api';
import { createVersionedContext, useVersionedContext } from '@backstage/version-bridge';
import 'react';

function getEntityRelations(entity, relationType, filter) {
  var _a;
  let entityNames = ((_a = entity == null ? void 0 : entity.relations) == null ? void 0 : _a.filter((r) => r.type === relationType).map((r) => parseEntityRef(r.targetRef))) || [];
  if (filter == null ? void 0 : filter.kind) {
    entityNames = entityNames.filter(
      (e) => e.kind.toLocaleLowerCase("en-US") === filter.kind.toLocaleLowerCase("en-US")
    );
  }
  return entityNames;
}

function isOwnerOf(owner, entity) {
  const possibleOwners = new Set(
    [
      ...getEntityRelations(owner, RELATION_MEMBER_OF, { kind: "group" }),
      ...owner ? [getCompoundEntityRef(owner)] : []
    ].map(stringifyEntityRef)
  );
  const owners = getEntityRelations(entity, RELATION_OWNED_BY).map(
    stringifyEntityRef
  );
  for (const ownerItem of owners) {
    if (possibleOwners.has(ownerItem)) {
      return true;
    }
  }
  return false;
}

createVersionedContext(
  "entity-context"
);
function useAsyncEntity() {
  const versionedHolder = useVersionedContext(
    "entity-context"
  );
  if (!versionedHolder) {
    throw new Error("Entity context is not available");
  }
  const value = versionedHolder.atVersion(1);
  if (!value) {
    throw new Error("EntityContext v1 not available");
  }
  const { entity, loading, error, refresh } = value;
  return { entity, loading, error, refresh };
}

function useEntityPermission(permission) {
  const {
    entity,
    loading: loadingEntity,
    error: entityError
  } = useAsyncEntity();
  const {
    allowed,
    loading: loadingPermission,
    error: permissionError
  } = usePermission({
    permission,
    resourceRef: entity ? stringifyEntityRef(entity) : void 0
  });
  if (loadingEntity || loadingPermission) {
    return { loading: true, allowed: false };
  }
  if (entityError) {
    return { loading: false, allowed: false, error: entityError };
  }
  return { loading: false, allowed, error: permissionError };
}

export { isOwnerOf, useEntityPermission };
